Compression changes before and after restructuring the whole project

Alxus228 opened the pull request:
In this pull request you can see _step by step_ code I've committed during adding **tests**, implementing **OOP**, writing **documentation** and fixing **bugs**.

Test coverage: 
![image](https://user-images.githubusercontent.com/40440883/160099952-5b249e6f-170b-4718-957c-5c0c9cbca307.png)

Commit:
Initial commit

Commit:
main commit

Commit:
main commit

Commit:
main commit

Commit:
Adding tests

Commit:
Adding tests

Commit:
Adding tests

Commit:
Adding tests

Commit:
Adding command structure and +,- fully and '>', '<', '.' operations partially

Commit:
Adding main implementation

Commit:
Adding tests and commands

Commit:
Adding input operation func()

Commit:
Adding input operation func()

Commit:
Adding zero operation func()
       copy operation func()
       paste operation func()

Commit:
Adding zero operation func()
       copy operation func()
       paste operation func()

Commit:
Adding loop operation func()

Commit:
Bug fix:
exucute operation in command interface doesn't need *memmory argument anymore

Commit:
Merging commands pull request

Adding Commands

Commit:
Bug fix:
exucute operation in command interface doesn't need *memmory argument anymore

Commit:
Fixing interface implementation bug
functions were implemented as pointers methods

Commit:
Finishing the implementation
implementing compiler and creating function Brainfuck where interpretate() and compile() are called
also changing naming in brainfuck_test.go and main.go

Commit:
Bug fix 1.0

Commit:
Bug fix 1.1:
Deleted "commands" unused variable
Added repeat = false in loopCLBOperation

Commit:
Renaming go module

Commit:
Renaming go module

Commit:
Renaming go module

Commit:
Bug fix and Debug:
Added check operation to check borders of loop before it starts
Renamed loopCheckLoopBordersOperation to loopCheckBordersOperation
Added debugging of compile() and interpretate()
Changed main input to "+++>+++++[-<+>].<." in order to understand why loops do not work

Commit:
Bug fix 1.2:
Added operation implementattion in loopCheckBordersOperation

Commit:
Bug fix 1.2:
Added operation implementattion in loopCheckBordersOperation
Some debug changes

Commit:
Bug fix 1.3:
Fixed runtime error 2 bug

Commit:
Bug fix 1.4:
Deleted debug

Commit:
Bug fix 1.5:
mem *memmory is a part of method argument now, not a part of the structure
debug added

Commit:
Bug fix 1.7:
Reworked function loopOperation.execute(mem *memmory); now it works correctly.
Deleted debug
Instead of the type loopCheckBordersOperation was added endLoopOperation : operation; this type's execute() command do nothing because it's only needed to help the interpretator append loopOperation into currentLoop.

Commit:
Bug fix 1.7:
Reworked function loopOperation.execute(mem *memmory); now it works correctly.
Deleted debug
Instead of the type loopCheckBordersOperation was added endLoopOperation : operation; this type's execute() command do nothing because it's only needed to help the interpretator append loopOperation into currentLoop.

Commit:
Bug fix 1.8:
operation type is now meets the requirements of the command interface

Commit:
Unneeded abstraction deleted
Type operation deleted and everything is working perfectly. It was just an extra layer of abstraction.
Also, added documentation for command interface.

Commit:
Documentation and Naming
Added documentation to all the types in brainfuck package
Fixed naming of the interpretate() and all the operations types were renamed in order to match the Go assigment.
Also, bool repeat variable was deleted from the loop structure because of its uselessness.

Commit:
Full Documentation
Evey type and function in brainfuck package has a documentation now.
Also added comments where the code isn't obvious.

Commit:
Merge pull request from commands

Documentation+
Naming+
All bugs fixed+
Abstraction+
Tests -

Commit:
Small changes2
copyPasteAccumulator moved to the command.go file.

Commit:
Small changes
copyPasteAccumulator moved to the command.go file.

Commit:
Merge pull request from commands

Small style changes

Commit:
Create README.md

Commit:
Added increment.execute() test

Commit:
Added '+' '-' '>' '<' execute() tests
Added protection against cellsSize

Commit:
Merge pull request #6 from Alxus228/tests

Tests

Commit:
Delete Brainfuck.iml

Commit:
Delete .idea directory

VitaliyShevchenko commented:
pls name this PR with a really good name and add a PR description, attach code coverage and results of manual testing.

VitaliyShevchenko commented on .idea/.gitignore:
```diff
@@ -1,8 +0,0 @@
-# Default ignored files
```
are you going to add .gitignore?

Alxus228 replied:
@VitaliyShevchenko Everything is done

Alxus228 replied:
> like your solution! Nice job!
Thank you!

@VitaliyShevchenko, could you please do another code review?

Alxus228 commented:
> pls name this PR with a really good name and add a PR description, attach code coverage and results of manual testing.
Updated

VitaliyShevchenko commented on main.go:
```diff
@@ -1,9 +1,9 @@
 
 func main() {
-	brainfuck.Interpetate("++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.")
+	brainfuck.Brainfuck("++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.c>p.")
```
`brainfuck` should DO something. Pls use verb instead of `Brainfuck`

Alxus228 replied:
Done

VitaliyShevchenko commented on brainfuck/command.go:
```diff
@@ -0,0 +1,92 @@
+package brainfuck
```
pls add tests for all commands

Alxus228 replied:
Done

Commit:
Added Interpret() instead Brainfuck()

Interpret() is now the function that can be called from API

Commit:
Changed the call function to Interpret()

Commit:
Renamed TestBrainfuck and added zero command tests

Commit:
Added copy and paste tests

Commit:
Added output tests and fixed copy paste tests

Commit:
Added loop tests

Commit:
Create .gitignore

VitaliyShevchenko commented:
pls add a doc for `brainfuck` package

VitaliyShevchenko commented on brainfuck/brainfuck.go:
```diff
@@ -1,66 +1,74 @@
-	//creating "main loop"
-	loopsStack = append(loopsStack, *new(loop))
-	loopsStack[0].executing = true
+// This function interpretates and compiles brainfuck code.
```
pls read how to write docs in Golang and fix docs.

Alxus228 replied:
Done, and also added package doc

VitaliyShevchenko commented on brainfuck/brainfuck.go:
```diff
@@ -1,66 +1,74 @@
+// This function interpretates and compiles brainfuck code.
+//
+// Specification of the language can be found on Wiki: https://en.wikipedia.org/wiki/Brainfuck.
+func Interpret(code string) {
```
it's better to move public funcs at the top of the file. So anyone else who wants to check what's inside this file will be able to see public API once the file is opened.

Alxus228 replied:
Done

VitaliyShevchenko commented on brainfuck/brainfuck.go:
```diff
@@ -1,66 +1,74 @@
+//
+// After the interpretation we will execute it, but not with the
+// currentLoop[0].execute(memmorySet), because we don't need to repeat it
+var currentLoop = []loop{
```
why are you using global var? I think it's not suitable in this case. It's better to return `currentLoop` and then pass it to `compile(currentLoop)`

VitaliyShevchenko commented on brainfuck/command.go:
```diff
@@ -0,0 +1,92 @@
+func (com endLoop) execute(mem *memmory) {}
+
+// This variable is needed for copy and paste commands.
+var copyPasteAccumulator byte
```
instead of having global variable I would move this var under the appropriate structs and use share it between them..

Alxus228 replied:
I don't understand, how to implement that without using any global variables, or without adding external code into compile() func

VitaliyShevchenko commented on brainfuck/memmory.go:
```diff
@@ -1,6 +1,11 @@
+// Type memmory mimics an actual memmory's functional but has much fewer capaticy.
+//
+// You can set any capaticy to the cells array you want if initial size isn't enough for your needs.
 type memmory struct {
```
memory

Alxus228 replied:
Done

VitaliyShevchenko commented on brainfuck/brainfuck.go:
```diff
@@ -1,66 +1,74 @@
+
+// Function compile execute all commands in the main loop.
+func compile() {
+	for _, com := range currentLoop[0].innerLoop {
```
pls create memory object before the range and pass it to each execute func
no need to use global vars

Alxus228 replied:
Done

VitaliyShevchenko commented on brainfuck/brainfuck_test.go:
```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+func TestExecute(t *testing.T) {
```
it's test for command.go. Pls create `command_test.go` and move it there

Alxus228 replied:
Done

VitaliyShevchenko commented on brainfuck/brainfuck_test.go:
```diff
@@ -0,0 +1,152 @@
+package brainfuck
```
here you should test your public API + some stuff only for `brainfuck.go` file. Everything else should be in the test files for appropriate functionality.

VitaliyShevchenko commented on brainfuck/brainfuck_test.go:
```diff
@@ -0,0 +1,152 @@
+	out, _ := ioutil.ReadAll(r)
+	os.Stdout = rescueStdout
+
+	if out[0] != 16 {
```
why you compare only count of symbols? 
Pls check the content as well

Alxus228 replied:
I tried to debug this piece of code, and it checks the content correctly.
Maybe, I haven't understood your advice, but `out` is a slice of `byte[]` and, during the debug, the only representation of this variable I received, was `[16]`

This is the proof:
![12](https://user-images.githubusercontent.com/40440883/161162993-eb79bb8d-5b89-4e14-9b28-8cc1ec6b56c2.jpg)

VitaliyShevchenko replied:
yes. that's ok.

VitaliyShevchenko commented:
pls read this: https://sparkbox.com/foundry/github_pull_requests_for_everyone

VitaliyShevchenko commented on README.md:
```diff
@@ -0,0 +1,8 @@
+# Brainfuck Interpreter
```
you can also add an example how to get your library, how to run the tests

Alxus228 replied:
Done

VitaliyShevchenko commented:
like your solution! Nice job!

Commit:
Renaming 'memmory' structure and restyle the code

Renamed 'memmory' to 'memory' and moved public API onto the top

Commit:
Delete memmory.go

Deleting memmory.go file because it was renamed to gramatically correct version "memory.go"

Commit:
Making memorySet and currentLoop inner variables

Commit:
Fixed grammatical errors and moved memorySet away

Fixed grammatical errors in documentation and moved creating global memory variable 'memorySet' into the compile() function

Commit:
Moving testing of command.go into appropriate file

Commit:
Adding the package documentation and standardizing

Added the brainfuck package documentation
Fixed documentation of Interpret function, so it meets the requirements of golang

Commit:
Adding installation and testing guide

Adding the description of how brainfuck package can be tested and installed.

Alxus228 commented:
>like your solution! Nice job!

Thank you!

@VitaliyShevchenko, could you please do another code review?

VitaliyShevchenko commented:
pls name the PR properly :)

AlryC committed:
Update README.md

Correcting grammar mistakes

AlryC committed:
Correct grammar mistakes

AlryC committed:
Correct grammar mistakes

AlryC committed:
Correct grammar mistakes

AlryC committed:
Update README.md

AlryC committed:
Update owner

AlryC committed:
Add Brainfuck in bash

AlryC committed:
Actually add bf in bash

AlryC committed:
~
